File: python/src/bsb/services_controller.py
from __future__ import annotations

from dataclasses import dataclass, field
from graphlib import TopologicalSorter
from typing import Any

from .base import BSBService, PluginCtor, dispose_all, validate_plugin_config
from .events_controller import SBEvents
from .observable import ObservableBackend
from .plugin_loader import LoadedPlugin, SBPlugins


@dataclass(slots=True)
class _SortedService:
    src_plugin_name: str
    plugin_name: str
    init_before_plugins: list[str]
    init_after_plugins: list[str]
    run_before_plugins: list[str]
    run_after_plugins: list[str]
    reference: BSBService
    clients: list[Any] = field(default_factory=list)


class SBServices:
    def __init__(
        self,
        app_id: str,
        mode: str,
        cwd: str,
        sb_plugins: SBPlugins,
        sb_events: SBEvents,
        observable_backend: ObservableBackend,
    ) -> None:
        self.app_id = app_id
        self.mode = mode
        self.cwd = cwd
        self.sb_plugins = sb_plugins
        self.sb_events = sb_events
        self.obs = observable_backend
        self._active_services: list[BSBService] = []

    async def dispose(self) -> None:
        await dispose_all(reversed(self._active_services))

    async def setup(self, sb_config: Any) -> None:
        plugins = await sb_config.get_service_plugins()
        if not plugins:
            raise RuntimeError("At least one enabled service is required")
        self._definitions = await sb_config.get_service_references()
        self.sb_events.set_services(self._definitions)
        for alias, plugin_def in plugins.items():
            await self._add_service(
                sb_config,
                alias,
                self._field(plugin_def, "plugin"),
                self._field(plugin_def, "package"),
                self._field(plugin_def, "version"),
                self._field(plugin_def, "language"),
            )
        for service in self._active_services:
            await self._remap_deps(sb_config, service)

    async def _remap_deps(self, sb_config: Any, service: BSBService) -> None:
        service.init_before_plugins = await self._map_plugins(sb_config, service.plugin_name, service.init_before_plugins)
        service.init_after_plugins = await self._map_plugins(sb_config, service.plugin_name, service.init_after_plugins)
        service.run_before_plugins = await self._map_plugins(sb_config, service.plugin_name, service.run_before_plugins)
        service.run_after_plugins = await self._map_plugins(sb_config, service.plugin_name, service.run_after_plugins)

    async def _map_plugins(self, sb_config: Any, ref_name: str, source: list[str]) -> list[str]:
        out: list[str] = []
        for plugin in source or []:
            if plugin in self._definitions:
                out.append(plugin)
            else:
                out.extend(alias for alias, definition in self._definitions.items() if definition.get("plugin") == plugin)
        return out

    async def _add_service(self, sb_config: Any, alias: str, plugin_ref: str, package: str | None, version=None, language=None) -> None:
        loaded: LoadedPlugin = await self.sb_plugins.load_plugin("service", package, plugin_ref or alias, alias, version, language)
        plugin_config = validate_plugin_config(
            loaded.service_config,
            await sb_config.get_plugin_config("service", alias),
            f"Invalid config for service {alias}",
        )
        service = loaded.plugin(
            PluginCtor(
                app_id=self.app_id,
                mode=self.mode,
                plugin_name=loaded.name,
                cwd=self.cwd,
                package_cwd=loaded.package_cwd,
                plugin_cwd=loaded.plugin_cwd,
                config=plugin_config,
                plugin_version=loaded.version,
                observable_backend=self.obs.for_plugin(alias),
                events=self.sb_events,
            )
        )
        self._active_services.append(service)

    @staticmethod
    def _field(defn: Any, key: str) -> Any:
        if isinstance(defn, dict):
            return defn.get(key)
        return getattr(defn, key, None)

    async def init(self) -> None:
        await self._sort_and_run_or_init("init")

    async def run(self) -> None:
        await self._sort_and_run_or_init("run")

    def _gather_list(self) -> list[_SortedService]:
        out: list[_SortedService] = []
        for service in self._active_services:
            out.append(
                _SortedService(
                    src_plugin_name=service.plugin_name,
                    plugin_name=service.plugin_name,
                    init_before_plugins=list(service.init_before_plugins or []),
                    init_after_plugins=list(service.init_after_plugins or []),
                    run_before_plugins=list(service.run_before_plugins or []),
                    run_after_plugins=list(service.run_after_plugins or []),
                    reference=service,
                )
            )
        return out

    def _sort_by_deps(self, phase: str, plugins: list[_SortedService]) -> list[_SortedService]:
        by_name = {plugin.plugin_name: plugin for plugin in plugins}
        dependencies: dict[str, set[str]] = {name: set() for name in by_name}
        for plugin in plugins:
            before = plugin.init_before_plugins if phase == "init" else plugin.run_before_plugins
            after = plugin.init_after_plugins if phase == "init" else plugin.run_after_plugins
            dependencies[plugin.plugin_name].update(name for name in after if name in by_name)
            for name in before:
                if name in dependencies:
                    dependencies[name].add(plugin.plugin_name)
        return [by_name[name] for name in TopologicalSorter(dependencies).static_order()]

    async def _sort_and_run_or_init(self, phase: str) -> None:
        plugins = self._sort_by_deps(phase, self._gather_list())
        if phase == "init":
            self._active_services = [plugin.reference for plugin in plugins]
        for plugin in plugins:
            trace = plugin.reference.create_trace(phase)
            fn = getattr(plugin.reference, phase)
            try:
                await fn(trace)
            except Exception as error:
                trace.error(error)
                raise
            finally:
                trace.end()

File: go/bsb/services_controller.go
package bsb

import (
	"context"
	"fmt"
	"sort"
)

// sortedService holds a service plugin with its dependency metadata for ordering.
type sortedService struct {
	name              string
	plugin            ServicePlugin
	initBeforePlugins []string
	initAfterPlugins  []string
	runBeforePlugins  []string
	runAfterPlugins   []string
}

// ServicesController manages service plugin instances with dependency ordering.
type ServicesController struct {
	backend  *ObservableBackend
	registry *PluginRegistry
	opts     BSBOptions
	services []*sortedService
}

// Init loads and initializes service plugins from config.
func (sc *ServicesController) Init(ctx context.Context, obs Observable, config *ConfigController, events *EventsController, backend *ObservableBackend) error {
	obs.Log().Info("loading service plugins")

	pluginDefs, err := config.GetServicePlugins(ctx, obs)
	if err != nil {
		obs.Log().Warn("no service plugins in config", map[string]any{
			"error": err.Error(),
		})
		return nil
	}

	// Load all service plugins
	for name, def := range pluginDefs {
		if !def.Enabled {
			obs.Log().Debug("skipping disabled service", map[string]any{"plugin": name})
			continue
		}

		pluginName := def.Plugin
		if pluginName == "" {
			pluginName = name
		}

		if !sc.registry.HasPlugin(PluginTypeService, pluginName) {
			obs.Log().Warn("service plugin not registered, skipping", map[string]any{"plugin": pluginName})
			continue
		}

		pluginConfig, err := config.GetPluginConfig(ctx, obs, PluginTypeService, name)
		if err != nil {
			obs.Log().Warn("failed to get service plugin config, using definition config", map[string]any{
				"plugin": pluginName,
				"error":  err.Error(),
			})
			pluginConfig = def.Config
		}

		plugin, err := sc.registry.CreateService(pluginName, pluginConfig)
		if err != nil {
			return fmt.Errorf("failed to create service plugin %q: %w", pluginName, err)
		}

		// Wire up the events facade
		meta := plugin.Metadata()
		resource := BuildResourceContext(meta.Name, meta.Version, sc.opts.AppID, sc.opts.Mode, sc.opts.Region)
		pe := NewPluginEvents(name, events.Primary(), backend, resource, NewEventSchemas())
		plugin.SetEvents(pe)
		plugin.SetObservableBackend(backend)

		sc.services = append(sc.services, &sortedService{
			name:              name,
			plugin:            plugin,
			initBeforePlugins: meta.InitBeforePlugins,
			initAfterPlugins:  meta.InitAfterPlugins,
			runBeforePlugins:  meta.RunBeforePlugins,
			runAfterPlugins:   meta.RunAfterPlugins,
		})

		obs.Log().Info("service plugin loaded", map[string]any{"plugin": name})
	}

	// Sort by init dependencies
	sorted, err := topologicalSort(sc.services, true)
	if err != nil {
		return fmt.Errorf("failed to sort service init order: %w", err)
	}
	sc.services = sorted

	// Initialize in dependency order
	for _, svc := range sc.services {
		meta := svc.plugin.Metadata()
		resource := BuildResourceContext(meta.Name, meta.Version, sc.opts.AppID, sc.opts.Mode, sc.opts.Region)
		svcObs := backend.CreateObservable(NewDTrace(), svc.name, resource)

		obs.Log().Debug("initializing service", map[string]any{"plugin": svc.name})
		if err := svc.plugin.Init(ctx, svcObs); err != nil {
			return fmt.Errorf("service %q init failed: %w", svc.name, err)
		}
	}

	obs.Log().Info("all service plugins initialized", map[string]any{
		"count": len(sc.services),
	})
	return nil
}

// Run starts all service plugins in dependency order.
func (sc *ServicesController) Run(ctx context.Context, obs Observable) error {
	// Re-sort by run dependencies
	sorted, err := topologicalSort(sc.services, false)
	if err != nil {
		return fmt.Errorf("failed to sort service run order: %w", err)
	}

	for _, svc := range sorted {
		meta := svc.plugin.Metadata()
		resource := BuildResourceContext(meta.Name, meta.Version, sc.opts.AppID, sc.opts.Mode, sc.opts.Region)
		svcObs := sc.backend.CreateObservable(NewDTrace(), svc.name, resource)

		obs.Log().Debug("running service", map[string]any{"plugin": svc.name})
		if err := svc.plugin.Run(ctx, svcObs); err != nil {
			return fmt.Errorf("service %q run failed: %w", svc.name, err)
		}
	}

	obs.Log().Info("all service plugins running", map[string]any{
		"count": len(sorted),
	})
	return nil
}

// Dispose cleans up all service plugins in reverse order.
func (sc *ServicesController) Dispose() error {
	var errs []error
	for i := len(sc.services) - 1; i >= 0; i-- {
		if err := sc.services[i].plugin.Dispose(); err != nil {
			errs = append(errs, fmt.Errorf("service %q dispose: %w", sc.services[i].name, err))
		}
	}
	if len(errs) > 0 {
		return fmt.Errorf("services dispose errors: %v", errs)
	}
	return nil
}

// topologicalSort orders services based on their before/after dependency declarations.
// If forInit is true, uses initBefore/initAfter; otherwise uses runBefore/runAfter.
func topologicalSort(services []*sortedService, forInit bool) ([]*sortedService, error) {
	if len(services) == 0 {
		return services, nil
	}

	// Build name -> index map
	nameIndex := make(map[string]int, len(services))
	for i, svc := range services {
		nameIndex[svc.name] = i
	}

	// Build adjacency: edges[a] = {b} means a must come before b
	n := len(services)
	inDegree := make([]int, n)
	adj := make([][]int, n)
	for i := range adj {
		adj[i] = []int{}
	}

	addEdge := func(from, to int) {
		adj[from] = append(adj[from], to)
		inDegree[to]++
	}

	for i, svc := range services {
		var before, after []string
		if forInit {
			before = svc.initBeforePlugins
			after = svc.initAfterPlugins
		} else {
			before = svc.runBeforePlugins
			after = svc.runAfterPlugins
		}

		// "before" means this plugin must init BEFORE those plugins
		for _, dep := range before {
			if j, ok := nameIndex[dep]; ok {
				addEdge(i, j)
			}
		}

		// "after" means this plugin must init AFTER those plugins
		for _, dep := range after {
			if j, ok := nameIndex[dep]; ok {
				addEdge(j, i)
			}
		}
	}

	// Kahn's algorithm
	queue := make([]int, 0)
	for i, deg := range inDegree {
		if deg == 0 {
			queue = append(queue, i)
		}
	}

	// Sort queue for deterministic output
	sort.Ints(queue)

	var result []*sortedService
	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		result = append(result, services[curr])

		for _, next := range adj[curr] {
			inDegree[next]--
			if inDegree[next] == 0 {
				queue = append(queue, next)
				sort.Ints(queue)
			}
		}
	}

	if len(result) != n {
		// Cycle detected -- find the involved plugins
		var cycled []string
		for i, deg := range inDegree {
			if deg > 0 {
				cycled = append(cycled, services[i].name)
			}
		}
		return nil, &DependencyCycleError{Plugins: cycled}
	}

	return result, nil
}
